lib/srv/discovery/fetchers/gke.go

```diff
@@ -0,0 +1,158 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fetchers
+
+import (
+	"context"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/cloud/gcp"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/trace"
+
+	"github.com/sirupsen/logrus"
+	containerpb "google.golang.org/genproto/googleapis/container/v1"
+)
+
+// GKEFetcherConfig configures the GKE fetcher.
+type GKEFetcherConfig struct {
+	// Client is the GCP GKE client.
+	Client gcp.GKEClient
+	// ProjectID is the projectID the cluster should belong to.
+	ProjectID string
+	// Location is the region where the clusters should be located.
+	// Wildcard "*" is supported.
+	Location string
+	// FilterLabels are the filter criteria.
+	FilterLabels types.Labels
+	// Log is the logger.
+	Log logrus.FieldLogger
+}
+
+// CheckAndSetDefaults validates and sets the defaults values.
+func (c *GKEFetcherConfig) CheckAndSetDefaults() error {
+	if c.Client == nil {
+		return trace.BadParameter("missing Client field")
+	}
+	if len(c.Location) == 0 {
+		return trace.BadParameter("missing Location field")
+	}
+
+	if len(c.FilterLabels) == 0 {
+		return trace.BadParameter("missing FilterLabels field")
+	}
+
+	if c.Log == nil {
+		c.Log = logrus.WithField(trace.Component, "fetcher:gke")
+	}
+	return nil
+}
+
+// gkeFetcher is a GKE fetcher.
+type gkeFetcher struct {
+	GKEFetcherConfig
+}
+
+// NewGKEFetcher creates a new GKE fetcher configuration.
+func NewGKEFetcher(cfg GKEFetcherConfig) (Fetcher, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &gkeFetcher{cfg}, nil
+}
+
+func (a *gkeFetcher) Get(ctx context.Context) (types.ResourcesWithLabels, error) {
+	clusters, err := a.getGKEClusters(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return clusters.AsResources(), nil
+}
+
+func (a *gkeFetcher) getGKEClusters(ctx context.Context) (types.KubeClusters, error) {
+	var (
+		clusters types.KubeClusters
+	)
+
+	gkeClusters, err := a.Client.ListClusters(ctx, a.ProjectID, a.Location)
+	for _, gkeCluster := range gkeClusters {
+		logger := a.Log.WithField("cluster_name", gkeCluster.Name)
+		cluster, err := a.getMatchingKubeCluster(gkeCluster)
+		// trace.CompareFailed is returned if the cluster did not match the matcher filtering labels
+		// or if the cluster is not yet active.
+		if trace.IsCompareFailed(err) {
+			logger.WithError(err).Debugf("Cluster %q did not match the filtering criteria.", gkeCluster.Name)
```

Comment:
It seems a bit odd to mix structured logging (e.g WithField) with a string that contains variables.